File: app/models/api/agent.py
# ruff: noqa: F401
from __future__ import annotations

from enum import StrEnum
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.constants import TWEET_SUGGESTION_MODEL
from app.models.api.base import UTCDateTime
from app.models.api.onboarding import OnboardingSelectedAggregator
from app.models.api.pagination import PaginationMetadata
from app.models.contracts import (
    ContentClassification,
    ContentStatus,
    ContentType,
    SummaryKind,
    SummaryVersion,
)


class AgentLibraryDocumentResponse(BaseModel):
    """One manifest entry for a personal markdown document."""

    relative_path: str
    content_id: int
    variant: Literal["source", "summary"]
    updated_at: UTCDateTime | None = None
    size_bytes: int
    checksum_sha256: str


class AgentLibraryManifestResponse(BaseModel):
    """Manifest of markdown documents available for CLI sync."""

    generated_at: UTCDateTime
    include_source: bool = True
    documents: list[AgentLibraryDocumentResponse]


class AgentLibraryFileResponse(BaseModel):
    """One markdown document payload for CLI sync download."""

    relative_path: str
    content_id: int
    variant: Literal["source", "summary"]
    updated_at: UTCDateTime | None = None
    checksum_sha256: str
    text: str


class AgentSearchRequest(BaseModel):
    """Machine-oriented external search request."""

    query: str = Field(..., min_length=2, max_length=200)
    limit: int = Field(default=10, ge=1, le=25)
    include_podcasts: bool = True


class AgentSearchResultResponse(BaseModel):
    """One agent search result."""

    kind: Literal["web", "podcast"]
    title: str
    url: str
    snippet: str | None = None
    source: str | None = None
    provider: str | None = None
    feed_url: str | None = None
    published_at: str | None = None
    score: float | None = None


class AgentSearchResponse(BaseModel):
    """Machine-oriented external search response."""

    results: list[AgentSearchResultResponse] = Field(default_factory=list)


class AgentOnboardingStartRequest(BaseModel):
    """Simplified async onboarding start request."""

    brief: str = Field(..., min_length=1, max_length=4000)
    preferences: dict[str, Any] | None = None
    seed_urls: list[str] = Field(default_factory=list)
    seed_feeds: list[str] = Field(default_factory=list)


class AgentOnboardingStartResponse(BaseModel):
    """Simplified async onboarding start response."""

    run_id: int
    status: str
    job_id: int | None = None


class AgentOnboardingCompleteRequest(BaseModel):
    """Complete simplified agent onboarding."""

    accept_all: bool = False
    source_ids: list[int] = Field(default_factory=list)
    selected_subreddits: list[str] = Field(default_factory=list)
    selected_aggregators: list[OnboardingSelectedAggregator] = Field(default_factory=list)

File: rust/crates/newsly-contracts/src/agent.rs
use chrono::{DateTime, Utc};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AgentLibraryDocumentVariant {
    Source,
    Summary,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AgentSearchResultKind {
    Web,
    Podcast,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentLibraryDocumentResponse {
    pub relative_path: String,
    pub content_id: i64,
    pub variant: AgentLibraryDocumentVariant,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub updated_at: Option<DateTime<Utc>>,
    pub size_bytes: usize,
    pub checksum_sha256: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentLibraryManifestResponse {
    #[schemars(with = "String")]
    #[schema(value_type = String, format = DateTime)]
    pub generated_at: DateTime<Utc>,
    #[serde(default = "default_true")]
    pub include_source: bool,
    pub documents: Vec<AgentLibraryDocumentResponse>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentLibraryFileResponse {
    pub relative_path: String,
    pub content_id: i64,
    pub variant: AgentLibraryDocumentVariant,
    #[schemars(with = "Option<String>")]
    #[schema(value_type = Option<String>, format = DateTime)]
    pub updated_at: Option<DateTime<Utc>>,
    pub checksum_sha256: String,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentSearchRequest {
    #[schemars(length(min = 2, max = 200))]
    #[schema(min_length = 2, max_length = 200)]
    pub query: String,
    #[serde(default = "default_search_limit")]
    #[schemars(range(min = 1, max = 25))]
    #[schema(minimum = 1, maximum = 25, default = 10)]
    pub limit: usize,
    #[serde(default = "default_true")]
    pub include_podcasts: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentSearchResultResponse {
    pub kind: AgentSearchResultKind,
    pub title: String,
    pub url: String,
    pub snippet: Option<String>,
    pub source: Option<String>,
    pub provider: Option<String>,
    pub feed_url: Option<String>,
    pub published_at: Option<String>,
    pub score: Option<f64>,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize, Deserialize, JsonSchema, ToSchema)]
pub struct AgentSearchResponse {
    #[serde(default)]
    pub results: Vec<AgentSearchResultResponse>,
}

const fn default_true() -> bool {
    true
}

const fn default_search_limit() -> usize {
    10
}
